packages/@ember/-internals/routing/lib/services/router.ts

```diff
@@ -61,7 +65,16 @@ function cleanURL(url: string, rootURL: string) {
    @class RouterService
  */
 export default class RouterService extends Service {
-  _router!: EmberRouter;
+  get _router(): EmberRouter {
+    let router = this[ROUTER];
+    if (router !== undefined) {
+      return router;
+    }
+    const owner = getOwner(this) as any;
+    router = owner.lookup('router:main');
+    router.setupRouter();
+    return (this[ROUTER] = router);
+  }
```

Comment:
Maybe we could refactor to:

```suggestion
  get _router(): EmberRouter {
    let router = this[ROUTER];
    if (router !== undefined) {
      return router;
    }
    
    const owner = getOwner(this) as any;
    owner.setupRouter();
    
    return this[ROUTER] = owner.router;
  }
  
  constructor(owner) {
    super(owner);
    
    let { router } = owner;
    
    router.on('routeWillChange', () => { /* snip */ });
    router.on('routedidChange', () => { /* snip */ });
  }
```

This would require moving that `didSetupRouter` flag back to the `owner` though...

What do you think?

Reply:
Do we need `EMBER_FRAMEWORK_OBJECT_OWNER_ARGUMENT` if moving the event listeners from `init` to `constructor`?

Reply:
I'm not sure I follow what you mean.

Reply:
Ah sorry please ignore this comment \_(:з」∠)\_ I must have mistaken https://github.com/emberjs/rfcs/pull/451 with something else